cf-custom-resources/lib/wkld-cert-validator.js

```diff
@@ -0,0 +1,881 @@
+    }
+    return id;
+  };
+};
+
+const rootHostedZoneIDContext = () => {
+  let id;
+  return async () => {
+    if (!id) {
+      id = await hostedZoneIDByName(`${domainName}`);
+    }
+    return id;
+  };
+};
+
+let hostedZoneID = {
+  app: appHostedZoneIDContext(),
+  root: rootHostedZoneIDContext(),
+};
+
+/**
+ * Upload a CloudFormation response object to S3.
+ *
+ * @param {object} event the Lambda event payload received by the handler function
+ * @param {object} context the Lambda context received by the handler function
+ * @param {string} responseStatus the response status, either 'SUCCESS' or 'FAILED'
+ * @param {string} physicalResourceId CloudFormation physical resource ID
+ * @param {object} [responseData] arbitrary response data object
+ * @param {string} [reason] reason for failure, if any, to convey to the user
+ * @returns {Promise} Promise that is resolved on success, or rejected on connection error or HTTP error response
+ */
+function report(
+  event,
+  context,
+  responseStatus,
+  physicalResourceId,
+  responseData,
+  reason
+) {
+  return new Promise((resolve, reject) => {
+    const https = require("https");
+    const { URL } = require("url");
+
+    let reasonWithLogInfo = `${reason} (Log: ${context.logGroupName}/${context.logStreamName})`;
+    let responseBody = JSON.stringify({
+      Status: responseStatus,
+      Reason: reasonWithLogInfo,
+      PhysicalResourceId: physicalResourceId || context.logStreamName,
+      StackId: event.StackId,
+      RequestId: event.RequestId,
+      LogicalResourceId: event.LogicalResourceId,
+      Data: responseData,
+    });
+
+    const parsedUrl = new URL(event.ResponseURL);
+    const options = {
+      hostname: parsedUrl.hostname,
+      port: 443,
+      path: parsedUrl.pathname + parsedUrl.search,
+      method: "PUT",
+      headers: {
+        "Content-Type": "",
+        "Content-Length": responseBody.length,
+      },
+    };
+
+    https
+      .request(options)
+      .on("error", reject)
+      .on("response", (res) => {
+        res.resume();
+        if (res.statusCode >= 400) {
+          reject(new Error(`Error ${res.statusCode}: ${res.statusMessage}`));
+        } else {
+          resolve();
+        }
+      })
+      .end(responseBody, "utf8");
+  });
+}
+
+exports.handler = async function (event, context) {
+  // Destruct resource properties into local variables.
+  const props = event.ResourceProperties;
+  let { LoadBalancerDNS: loadBalancerDNS } = props;
+  const aliases = new Set(props.Aliases);
+
+  // Initialize global variables.
+  envHostedZoneID = props.EnvHostedZoneId;
+  envName = props.EnvName;
+  appName = props.AppName;
+  serviceName = props.ServiceName;
+  domainName = props.DomainName;
+  rootDNSRole = props.RootDNSRole;
+  cloudFrontCert = props.IsCloudFrontCertificate;
+  certificateDomain = cloudFrontCert
+    ? `${serviceName}.${envName}.${appName}.${domainName}`
+    : `${serviceName}-nlb.${envName}.${appName}.${domainName}`;
```

Comment:
hm I don't recall why the NLB one has the `-nlb` at the end. Perhaps it was the default NLB alias, and that the default alias can't collapse with the default alias for HTTPS? Would static site have the same issue?

Reply:
it was the default NLB alias. I don't think it'll be an issue because the alias will be dedicated to this static site service only. And unlike LBWS, if they don't use alias we won't have any default alias for them. Since we always create a CF by default we can just use the CF default domain and it'll be unique instead of creating a default alias for them.

Reply:
> And unlike LBWS, if they don't use alias we won't have any default alias for them. 

Should we write the CFN condition differently for static site then? Right now the condition triggers on "HasAssociatedDomain", whereas the intention is to trigger it when "there is an associated domain" + "there is a static site alias"

Reply:
In addition, does it matter if we use `${serviceName}.${envName}.${appName}.${domainName}` as the main domain for static site cert? Perhaps we should use the specific alias specified under `http.alias` for the static site instead?

Reply:
> Right now the condition triggers on "HasAssociatedDomain",
Right now it is also in the go template's if statement for if alias exists. So it think `whereas the intention is to trigger it when "there is an associated domain" + "there is a static site alias"` this is how it works now

Reply:
> This would validate ${serviceName}.${envName}.${appName}.${domainName} as well although they never asked for it. Not sure if this is an issue.

I feel it is probably fine because it'll make it easier for them to switch to the alias we reserve for them? Also it'll make the static site certificate consistent with the other Copilot managed service certificates.